### PR TITLE
Refactor/rate limiter abstractions

### PR DESCRIPTION
**Pull request by Garnek20**
In this PR you'll find:
- decoupling of `reverseproxy` and `scope` from scope's management responsibility (which was scattered across those 2 abstractions)

Why?
To simplify plugging `RateLimiter` and evolution of the `chproxy`

### CODE REVIEW

File: scope.go [hunk 1]
```diff
@@ -509,6 +512,12 @@ func (up usersProfile) newUser(u config.User) (*user, error) {
 		}
 	}
 
+	var rateLimiter limiter.RateLimiter
+	if u.ReqPerMin > 0 {
+		// todo: if passed info about rate limiter enable it
```
Comment (Garnek20) [inline scope.go:517]: updated in the separate PR #14

File: scope.go [hunk 2]
```diff
@@ -203,45 +209,42 @@ func (s *scope) inc() error {
...
-	// These races become innocent with the given check.
-	if s.user.reqPerMin > 0 && int32(uRPM) > 0 && uRPM > s.user.reqPerMin {
-		err = fmt.Errorf("rate limit for user %q is exceeded: requests_per_minute limit: %d",
-			s.user.name, s.user.reqPerMin)
 	}
-	if s.clusterUser.reqPerMin > 0 && int32(cRPM) > 0 && cRPM > s.clusterUser.reqPerMin {
-		err = fmt.Errorf("rate limit for cluster user %q is exceeded: requests_per_minute limit: %d",
-			s.clusterUser.name, s.clusterUser.reqPerMin)
+
+	if s.clusterUser.rateLimiter != nil && err == nil {
```
Comment (Garnek20) [inline scope.go:226]: updated in the separate PR #14

File: scope.go [hunk 3]
```diff
@@ -559,12 +569,20 @@ func newClusterUser(cu config.ClusterUser) *clusterUser {
 	if cu.MaxQueueSize > 0 {
 		queueCh = make(chan struct{}, cu.MaxQueueSize)
 	}
+
+	var rateLimiter limiter.RateLimiter
+	if cu.ReqPerMin > 0 {
```
Comment (Garnek20) [inline scope.go:574]: updated in the separate PR #14

File: scope_manager.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+	"fmt"
+	"github.com/Vertamedia/chproxy/config"
+	"github.com/Vertamedia/chproxy/log"
+	"github.com/Vertamedia/chproxy/services/cache"
+	"github.com/prometheus/client_golang/prometheus"
+	"net/http"
+	"sync"
+)
+
+type ConfigListener interface {
```
Comment (Garnek20) [inline scope_manager.go:13]: Those are abstractions defined to decouple existing code. Before scope management or state management was dispatched across different places and some of it (what was the most inconvenient for the future evolution) was coupled to `reverseproxy`

File: services/ratelimiter/fixed_window_rate_limiter.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package ratelimiter
+
+import (
+	"time"
+)
+
+type fixedWindowRateLimiter struct {
```
Comment (Garnek20) [inline services/ratelimiter/fixed_window_rate_limiter.go:7]: implementation of fixed window is just extracted out from 2 different places (yes, it was scattered) and slightly changed to be able to remove previously existing leak (run, was exposed outside)

File: services/ratelimiter/fixed_window_rate_limiter.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	}
+	return false
+}
+
+func (rl *fixedWindowRateLimiter) run(done <-chan struct{}) {
+	for {
+		select {
+		case <-done:
+			return
+		case <-time.After(time.Minute):
```
Comment (Garnek20) [inline services/ratelimiter/fixed_window_rate_limiter.go:37]: this is hardcoded here. It's updated in the PR #14

File: main_test.go [hunk 7]
```diff
@@ -97,7 +96,8 @@ func TestServe(t *testing.T) {
 					t.Fatalf("err while getting file %q info: %s", path, err)
 				}
 				rw := httptest.NewRecorder()
-				cc := proxy.caches["https_cache"]
+				scMan, _ := proxy.scopeManager.(*scopeManager)
```
Comment (r-zenine) [inline main_test.go:99]: Can you please catch the error and handle it ?

File: main_test.go [hunk 8]
```diff
@@ -142,7 +142,8 @@ func TestServe(t *testing.T) {
...
```
Comment (r-zenine) [inline main_test.go:145]: same here

File: scope_manager.go [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+
+type ErrWithStatusCode struct {
+	err        error
+	StatusCode int
+}
+
+type scopeManager struct {
+	// configLock serializes access to applyConfig.
+	// It protects reload* fields.
+	configLock sync.Mutex
```
Comment (r-zenine) [inline scope_manager.go:52]: you have to take this one by pointer, otherwise, your lock won't work

File: scope_manager.go [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+	reloadSignal chan struct{}
+	reloadWG     sync.WaitGroup
+
+	// lock protects users, clusters and caches.
+	// RWMutex enables concurrent access to getScope.
+	lock sync.RWMutex
```
Comment (r-zenine) [inline scope_manager.go:59]: same here

File: services/ratelimiter/fixed_window_rate_limiter.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+			rl.Store(0)
```
Comment (r-zenine) [inline services/ratelimiter/fixed_window_rate_limiter.go:38]: If a user does a burst of queries just after this store, are we going to block him ? I am not sure

Comment (Garnek20) [reply services/ratelimiter/fixed_window_rate_limiter.go:38]: No, it's the implementation of fixed window that was there. I just extracted it out

File: scope_manager.go (hunk 9 above)
Comment (Garnek20) [reply scope_manager.go:54]: We access to it via pointer receivers, therefore it should work. Am i missing something?

**Pull request merged**